File: LeetCode/Tree/Medium/98. Validate Binary Search Tree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/*
Intuition
1) recursion/DFS problem since I have to check for every node if bst propperty is present
2) all the the children node should have thier return 'and'since we want to check if its false
3) It can have one child also
4) if child not there return True
5) 

*/


class Solution {
public:

    bool dfs(TreeNode* root,long long max,long long min){
        if (root==NULL) return true;


        cout<<max;
        if (root->val > min and root->val<max){
            // if both subtree are bst parent will also be bst
            // updating max in left subtree(so that no element goes beyond parent parent node) and min in right subtree
            // for eg - 5 left-> 1 right-> 6  this shouldnt happen hence we keeping check with max value for left side
            // and 
            return dfs(root->left,root->val,min) and dfs(root->right,max,root->val);
        }

        return false;

    }


    bool isValidBST(TreeNode* root) {
        long long max = 2147483649;
        long long min = -2147483649;
        return dfs(root,max,min);
    }
};

File: LeetCode/Tree/Medium/98. Validate Binary Search Tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root):
        
        def helper(root,min_no,max_no):
            if not root:
                return True

            if min_no<root.val<max_no:
                return helper(root.left,min_no,root.val) and helper(root.right,root.val,max_no) 

            return False

        if not root:
            return True
        max_no = float(+inf)
        min_no = float(-inf) 
        return helper(root,min_no,max_no)

        